week7/answers.sql

```diff
@@ -1,3 +1,284 @@
+			INNER JOIN items ON items.item_id = equipped.item_id
+			INNER JOIN team_members tm ON c.character_id = tm.character_id
+			INNER JOIN teams t ON tm.team_id = t.team_id
+			ORDER BY item_name;
+
+
+DELIMITER ;;
+CREATE FUNCTION armor_total(character_id INT)
+RETURNS INT UNSIGNED
+READS SQL DATA
+BEGIN
+	DECLARE equipped_armor_sum INT;
+    DECLARE armor_stat INT;
+    SELECT 
+		SUM(items.armor) INTO equipped_armor_sum
+		FROM equipped e
+			INNER JOIN items ON e.item_id = items.item_id
+		WHERE 
+			e.character_id = character_id
+		GROUP BY e.character_id;
+	SELECT armor INTO armor_stat FROM character_stats cs
+		WHERE cs.character_id = character_id;
+		RETURN equipped_armor_sum + armor_stat;
+END;;
+DELIMITER ;
+
+DELIMITER ;;
+-- I got a little stumped here, so I looked up ways to save data into individual variables.
+-- Found this on stack overflow:
+-- https://stackoverflow.com/questions/4823880/sql-server-select-into-variable
+-- Then I looked at the CURSOR lesson again and realized I didn't need to do that, but
+-- I thought I should include the link anyway.
+CREATE PROCEDURE equip(selected_inventory_id INT)
+	BEGIN
+    DECLARE equipping_inventory_id INT;
+    DECLARE equipping_character_id INT;
+    DECLARE equipping_item_id INT;
+    
+	SELECT * INTO equipping_inventory_id, equipping_character_id, equipping_item_id
+			FROM inventory i 
+			WHERE i.inventory_id = selected_inventory_id;
+            
+	INSERT INTO 
+		equipped (character_id, item_id) 
+        VALUES
+			(equipping_character_id, equipping_item_id);
+	
+    DELETE FROM inventory WHERE inventory_id = selected_inventory_id;
+	END;;
+DELIMITER ;
+
+
+DELIMITER ;;
+CREATE PROCEDURE unequip(selected_equipped_id INT)
+	BEGIN
+    DECLARE unequipping_equipped_id INT;
+    DECLARE unequipping_character_id INT;
+    DECLARE unequipping_item_id INT;
+    
+	SELECT * INTO unequipping_equipped_id, unequipping_character_id, unequipping_item_id
+			FROM equipped 
+			WHERE equipped.equipped_id = selected_equipped_id;
+            
+	INSERT INTO 
+		inventory (character_id, item_id) 
+        VALUES
+			(unequipping_character_id, unequipping_item_id);
+	
+    DELETE FROM equipped WHERE equipped_id = selected_equipped_id;
+	END;;
+DELIMITER ;
+
+DELIMITER ;;
+CREATE PROCEDURE attack(id_of_character_being_attacked INT, id_of_equipped_item_used_for_attack INT)
+	BEGIN
+    DECLARE total_armor INT;
+    DECLARE item_damage INT;
+    DECLARE current_health INT;
+    SELECT armor_total(id_of_character_being_attacked) INTO total_armor;
+    
+    SELECT i.damage INTO item_damage 
+		FROM equipped e 
+			INNER JOIN items i ON e.item_id = i.item_id
+        WHERE e.equipped_id = id_of_equipped_item_used_for_attack;
+    
+    IF item_damage > total_armor THEN
+		SELECT health INTO current_health FROM character_stats cs 
+			WHERE cs.character_id = id_of_character_being_attacked;
+		SET current_health = current_health - (item_damage - total_armor);
+		UPDATE character_stats cs
```

Comment:
Only an optimization and not required: if you use the UPDATE in the ELSE clause of the IF statement you use for DELETE FROM, then you can save a call to UPDATE in cases where a character simply dies.

Reply:
Interesting, that gracefully handles character death without an extra trip to the database. I have now included it.